File: Script/K.py
from math import exp
from math import pi, e, cos, sin

# def fft(x, inv=False):
#     n = size(x)
#     j = 0
#     for i in range(n):
#         if (i < j): x[i], x[j] = x[j], x[i]
#         m = n >> 1
#         while (1 <= m and m <= j):
#             j -= m
#             m >>= 1
#         j += m
#     mx = 1
#     while(mx < n):
#         wp = exp(complex(0, (-1 if inv else 1) *  pi / mx))
#         w = 1
#         for
#
#
#         mx <<= 1


def fft(array, n, inv=False):
    if n == 1:
        return
    theta = (pi * 2) / n
    if inv:
        Wn = pow(e, (-2 * pi * 1j) / n)
    else:
        Wn = pow(e, (2 * pi * 1j) / n)

    W = 1

    a = [0] * (n // 2)
    b = [0] * (n // 2)

    for i in range(n):
        if (i & 1):
            b[i // 2] = array[i]
        else:
            a[i // 2] = array[i]

    fft(a, n // 2, inv)
    fft(b, n // 2, inv)

    for i in range(n // 2):
        array[i] = a[i] + W * b[i]
        array[i + n // 2] = a[i] - W * b[i]
        W *= Wn

def encode(n):
    map_dir = {}
    map_dir['n'] = 0
    map_dir['w'] = 1
    map_dir['e'] = 2
    map_dir['s'] = 3
    te = [0 for _ in range(n)]

    for i in range(n):
        s = input()
        if s == '?':
            continue
        dir, step = s.split()
        te[i] = int(step) * 4 + map_dir[dir]
    return te


n, m = map(int, input().split())
a, b = encode(n), encode(m)
size = 2 * n + 1
while (size & (size - 1)): size += 1
bits = [i for i in range(5)]
match = [True for _ in range(n - m + 1)]

for bit in bits:
    A = [complex(0,0) for _ in range(size)]
    B = [complex(0,0) for _ in range(size)]
    F = [0 for _ in range(n - m + 1)]
    G = [0 for _ in range(n - m + 1)]

    for i in range(n): A[n-i-1] = complex(0 if (a[i] == 0) else 1 if (a[i] & (1 << bit)) else -1, 0)
    for i in range(m): B[i] = complex(0 if (b[i] == 0) else 1 if (b[i] & (1 << bit)) else -1, 0)

    fft(A, size)
    fft(B, size)
    for i in range(size): A[i] *= B[i]
    fft(A, size, inv=True)
    for i in range(n - m + 1): F[i] = int(round(A[n-i-1].real))
    #print(F)

    A = [complex(0, 0) for _ in range(size)]
    B = [complex(0, 0) for _ in range(size)]
    for i in range(n): A[n - i - 1] = complex(0 if (a[i] == 0) else 1, 0)
    for i in range(m): B[i] = complex(0 if (b[i] == 0) else 1, 0)
    fft(A, size)
    fft(B, size)
    for i in range(size): A[i] *= B[i]
    fft(A, size, inv=True)
    for i in range(n - m + 1): G[i] = int(round(A[n - i - 1].real))
    #print(G)

    for i in range(n - m + 1): match[i] = match[i] and (F[i] == G[i])

print(sum(match))

File: Script/K.cpp
#include <bits/stdc++.h>
#include <complex>
using namespace std;
const double pi = 3.14159265358979323846;

void fft(complex<long double> *x, int n, bool inv=false) {
  for (int i = 0, j = 0; i < n; i++) {
    if (i < j) swap(x[i], x[j]);
    int m = n>>1;
    while (1 <= m && m <= j) j -= m, m >>= 1;
    j += m;
  }
  for (int mx = 1; mx < n; mx <<= 1) {
    complex<long double> wp, w = 1;
    if (inv) wp = exp(complex<long double>(0, pi / mx));
    else wp = exp(complex<long double>(0, -pi / mx));

    for (int m = 0; m < mx; m++, w *= wp) {
      for (int i = m; i < n; i += mx << 1) {
        complex<long double> t = x[i + mx] * w;
        x[i + mx] = x[i] - t;
        x[i] += t; }
    }
  }
  if (inv) for(int i = 0; i < n; i++) x[i] /= complex<long double>(n);
}

vector<int> input(int n) {
  map<char, int> map_dir;
  map_dir['n'] = 0;
  map_dir['w'] = 1;
  map_dir['e'] = 2;
  map_dir['s'] = 3;

  vector<int> te(n, 0);
  for(int i = 0; i < n; i++) {
    char s; cin >> s;
    if (s == '?') continue;

    int k; cin >> k;
    te[i] = k * 4 + map_dir[s];
  }
  return te;
}
int F[600000], G[600000];
bool H[600000];

int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);

  int n, m;
  cin >> n >> m;
  for (int i = 0; i < n - m + 1; i++) H[i] = true;
  vector<int> a = input(n), b = input(m);

  int size = 2 * n+1;
  while (size & (size-1)) size++;
  complex<long double> *A = new complex<long double>[size], *B = new complex<long double>[size];

  vector<int> bits = {0,1,2,3,4};
  for(auto bit: bits) {
    for(int i = 0; i < size; i++) A[i] = 0, B[i] = 0;
    for(int i = 0; i < n; i++) A[n-i-1] = (a[i]==0)? 0 : (a[i] & (1 << bit)? 1 : -1);
    for(int i = 0; i < m; i++) B[i] = (b[i]==0)? 0 : (b[i] & (1 << bit)? 1 : -1);
    fft(A, size);
    fft(B, size);
    for(int i = 0; i < size; i++) A[i] *= B[i];
    fft(A, size, true);
    for(int i = 0; i < n - m + 1; i++) F[i] = int(round(real(A[n-1-i])));

    for(int i = 0; i < size; i++) A[i] = 0, B[i] = 0;
    for(int i = 0; i < n; i++) A[n-i-1] = (a[i] == 0)? 0 : 1;
    for(int i = 0; i < m; i++) B[i] = (b[i] == 0)? 0 : 1;
    fft(A, size);
    fft(B, size);
    for(int i = 0; i < size; i++) A[i] *= B[i];
    fft(A, size, true);
    for(int i = 0; i < n - m + 1; i++) G[i] = int(round(real(A[n-1-i])));

    for(int i = 0; i < n - m + 1; i++) H[i] = H[i] && F[i] == G[i];
  }

  int res = 0;
  for(int i = 0; i < n - m + 1; i++) if (H[i]) res++;
  cout << res;
}

main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    int n, m; cin >> n >> m;
    for (int i = 0; i < n - m + 1; i++) H[i] = true;
    vector<int> a = input(n)
}